tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,258 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Fine Tune the base model."""
+
+from datetime import datetime
+import glob
+import os
+import time
+from typing import Any, List
+
+from absl import app
+from absl import flags
+from google.cloud import storage
+import gspread
+import pandas as pd
+from sentence_transformers import InputExample
+from sentence_transformers import losses
+from sentence_transformers import SentenceTransformer
+from torch.utils.data import DataLoader
+import utils
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('model_name_v2', 'all-MiniLM-L6-v2', 'Model name')
+flags.DEFINE_string('bucket_name_v2', 'datcom-nl-models', 'Storage bucket')
+flags.DEFINE_string('embeddings_size', '',
+                    'Valid values are: "small", "medium"')
+
+flags.DEFINE_string(
+    'autogen_input_basedir', 'data/autogen_input',
+    'Base path for CSVs with autogenerated SVs with name and description. '
+    'The actual path is `{--autogen_input_base}/{--embeddings_size}/*.csv`.')
+
+flags.DEFINE_string('alternatives_filepattern', 'data/alternatives/*.csv',
+                    'File pattern (relative) for CSVs with alternatives')
+
+flags.DEFINE_string('sentence_pairs_filepath',
+                    'data/finetuning/sentence_pairs.csv',
+                    'File path for sentence pairs CSV')
+
+BATCH_SIZE = 256
+NUM_WARMUP_STEPS = 10
+NUM_EPOCHS = 10
+
+VERY_HIGH_MATCH_SCORE = 0.95
+HIGH_MATCH_SCORE = 0.90
+MEDIUM_HIGH_MATCH_SCORE = 0.85
+
+
+def _upload_to_gcs(ctx: utils.Context,
+                   gcs_path: str,
+                   local_filepath: str,
+                   empty_folder: bool = False) -> None:
+  print(f'Uploading {local_filepath}')
+  blob = ctx.bucket.blob(gcs_path)
+
+  if empty_folder:
+    blob.upload_from_string(
+        '', content_type='application/x-www-form-urlencoded;charset=UTF-8')
+  else:
+    blob.upload_from_filename(local_filepath)
+  print(f'Path in GCS: {gcs_path}')
+
+
+def _make_gcs_model_folder(base_model_name: str) -> str:
+  now = datetime.now()
+
+  month_str = utils.two_digits(now.month)
+  day_str = utils.two_digits(now.day)
+  hour_str = utils.two_digits(now.hour)
+  minute_str = utils.two_digits(now.minute)
+  second_str = utils.two_digits(now.second)
+
+  return f"model_{base_model_name}_{now.year}_{month_str}_{day_str}_{hour_str}_{minute_str}_{second_str}"
+
+
+def _alternatives(autogen_input_filepattern: str,
+                  alternatives_filepattern: str) -> pd.DataFrame:
+
+  df_svs = pd.DataFrame()
+  # Append autogen CSVs if any.
+  autogen_dfs = []
+  for autogen_csv in sorted(glob.glob(autogen_input_filepattern)):
+    print(f'Processing autogen input file: {autogen_csv}')
+    autogen_dfs.append(pd.read_csv(autogen_csv).fillna(""))
+  if autogen_dfs:
+    df_svs = pd.concat([df_svs] + autogen_dfs)
```

Comment:
Is `[df_svs]` needed here given its empty?

Reply:
Good point. Removed.